plutus-contract/src/Plutus/Contract/Test.hs

```diff
@@ -330,6 +331,18 @@ dataAtAddress address check =
               <+> foldMap (foldMap pretty . Ledger.txData . Ledger.txOutTxTx) utxo)
       pure result
 
+outputsAtAddress :: forall d. FromData d => Address -> ([(d, Value)] -> Bool) -> TracePredicate
+outputsAtAddress address check =
+  flip postMapM (L.generalize $ Folds.utxoAtAddress address) $ \utxo -> do
+    let datums = mapMaybe (uncurry $ getTxOutDatum @d) $ M.toList utxo
```

Comment:
maybe there should be a better error condition if the datum is not present. Or else you could change the type of `outputsAtAddress` to

```haskell
outputsAtAddress :: forall d. FromData d => Address -> ([(Either DatumHash d, Value)] -> Bool) -> TracePredicate
```

Just so that users can be more explicit about whether they expect a datum to be present.

Reply:
I have updated the source code.